src/lib.rs

```diff
@@ -42,264 +46,186 @@ pub struct ShellCommandOutput {
 
 pub struct BevyLocalCommandsPlugin;
 
 impl Plugin for BevyLocalCommandsPlugin {
     fn build(&self, app: &mut App) {
-        app.add_event::<RunShellCommand>()
-            .add_event::<ShellCommandStarted>()
-            .add_event::<ShellCommandOutput>()
-            .add_event::<KillShellCommand>()
-            .add_event::<ShellCommandCompleted>()
-            .init_resource::<ActiveShellCommands>()
+        app.add_event::<RunProcess>()
+            .add_event::<ProcessStarted>()
+            .add_event::<ProcessOutputEvent>()
+            .add_event::<KillProcess>()
+            .add_event::<ProcessCompleted>()
+            .init_resource::<ActiveProcessMap>()
             .add_systems(
                 Update,
                 (
-                    handle_new_shell_commands,
-                    handle_shell_command_output,
-                    handle_kill_shell_command,
-                    handle_completed_shell_commands,
-                ),
+                    handle_new_process,
+                    handle_kill_process,
+                    handle_process_output,
+                    handle_completed_process,
+                )
+                    .chain(),
             );
     }
 }
 
-fn get_shell_command_string(run_shell_command: &RunShellCommand) -> String {
-    format!(
-        "{} {}",
-        run_shell_command.program,
-        &run_shell_command.arguments.join(" ")
-    )
-}
-
-fn handle_new_shell_commands(
-    mut run_shell_command_event: EventReader<RunShellCommand>,
-    mut shell_command_started_event: EventWriter<ShellCommandStarted>,
-    mut shell_command_completed_event: EventWriter<ShellCommandCompleted>,
-    mut active_shell_commands: ResMut<ActiveShellCommands>,
+fn handle_new_process(
+    mut run_process_event: EventReader<RunProcess>,
+    mut process_started_event: EventWriter<ProcessStarted>,
+    mut process_error_event: EventWriter<ProcessError>,
+    mut active_process_map: ResMut<ActiveProcessMap>,
 ) {
-    for run_shell_command in run_shell_command_event.read() {
-        let command_string = get_shell_command_string(run_shell_command);
-        let expression = cmd(&run_shell_command.program, &run_shell_command.arguments);
-        let active_shell_command = spawn_shell_command(command_string.clone(), expression);
-        if active_shell_command.is_none() {
-            shell_command_completed_event.send(ShellCommandCompleted {
-                pid: 0,
-                command: command_string.clone(),
-                success: false,
-                output_buffer: vec![],
-            });
+    for run_shell_command in run_process_event.read() {
+        // Assemble the command
+        let mut cmd = Command::new(run_shell_command.program.clone());
+        cmd.args(run_shell_command.arguments.clone())
+            .stdout(Stdio::piped());
+
+        let Ok(active_process) = spawn_process(cmd) else {
+            process_error_event.send(ProcessError::FailedToStart);
```

Comment:
Can we add the program/args to the event? Just so the user can reference which command failed to start.

Reply:
Good point.
For now, I added the command string, as the PID is not available yet.
Maybe in the future we can return something that's of more practical use.

Reply:
Perfect! Thanks @TimJentzsch